### PR TITLE
[cxx-interop] Create benchmarks for using std::span in Swift

### PR DESCRIPTION
**Pull request by susmonteiro**

### CODE REVIEW

**Comment by susmonteiro** [comment]
@swift-ci please benchmark

**Comment by susmonteiro** [comment]
@swift-ci please benchmark

**Comment by susmonteiro** [comment]
@swift-ci please benchmark

**Comment by susmonteiro** [comment]
@swift-ci please benchmark

**Comment by susmonteiro** [comment]
@swift-ci please smoke test

File: test/Interop/Cxx/stdlib/use-std-span.swift [hunk 1]
```diff
@@ -176,6 +176,18 @@ StdSpanTestSuite.test("Init SpanOfInt") {
   expectFalse(cs.empty())
 }
 
+StdSpanTestSuite.test("Init SpanOfInt from Swift array") {
```
Comment (egorzhdan) [inline test/Interop/Cxx/stdlib/use-std-span.swift:179]: Since these changes aren't benchmarks, could you please extract them to a separate PR?

Comment (susmonteiro) [reply test/Interop/Cxx/stdlib/use-std-span.swift:179]: Ok sure!

**Comment by susmonteiro** [comment]
@swift-ci please benchmark

**Comment by susmonteiro** [comment]
@swift-ci please smoke test

**Comment by susmonteiro** [comment]
@swift-ci please benchmark

**Comment by susmonteiro** [comment]
@swift-ci please smoke test

**Comment by susmonteiro** [comment]
@swift-ci please benchmark

**Comment by susmonteiro** [comment]
@swift-ci please smoke test

**Comment by susmonteiro** [comment]
@swift-ci please benchmark

File: benchmark/utils/CxxTests/CxxStdlibPerformance.h [hunk 2]
```diff
@@ -29,6 +44,18 @@ inline void initSet(size_t size) {
     }
 }
 
+#if !defined(__linux__)
```
Comment (egorzhdan) [inline benchmark/utils/CxxTests/CxxStdlibPerformance.h:47]: Could we use `#if __has_include(<span>)` instead of `!defined(__linux__)`? That would make the reasoning behind the guard slightly more clear, and it would be easier to run the benchmarks locally with newer Linuxes that have the `span` header installed.
(It's fine to leave the `#if os` check on the Swift side as is, since I don't think there is a better way to perform this check in Swift)

Comment (susmonteiro) [reply benchmark/utils/CxxTests/CxxStdlibPerformance.h:47]: Of course, thanks!

**Comment by susmonteiro** [comment]
@swift-ci please benchmark

**Comment by susmonteiro** [comment]
@swift-ci please benchmark

**Comment by susmonteiro** [comment]
@swift-ci please benchmark

**Comment by susmonteiro** [comment]
@swift-ci please benchmark

**Comment by susmonteiro** [comment]
@swift-ci please benchmark

**Comment by susmonteiro** [comment]
@swift-ci please benchmark

**Comment by susmonteiro** [comment]
@swift-ci please benchmark

**Comment by susmonteiro** [comment]
@swift-ci please smoke test

File: benchmark/cxx-source/CxxSpanTests.swift [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+      sum = sum &+ x
+    }
+  }
+  blackHole(sum)
+}
+
+@inline(never)
+public func run_CxxSpanOfU32_MapSpan(_ n: Int) {
+  for _ in 0..<(n * iterRepeatFactor) {
+    let result = span.map { $0 + 5 }
```
Comment (egorzhdan) [inline benchmark/cxx-source/CxxSpanTests.swift:99]: Should this `+` also have no overflow check?

Comment (susmonteiro) [reply benchmark/cxx-source/CxxSpanTests.swift:99]: Yes, thanks!

**Comment by susmonteiro** [comment]
@swift-ci please benchmark

**Comment by susmonteiro** [comment]
@swift-ci please smoke test

**Review by egorzhdan — APPROVED** [review]
Thanks!

**Review by eeckstein — APPROVED** [review]

**Comment by tshortli** [comment]
I think this PR may have broken the macOS toolchain build, could you take a look?

```
/Users/ec2-user/jenkins/workspace/swift-PR-toolchain-macos/branch-main/swift/benchmark/utils/CxxTests/CxxStdlibPerformance.h:18:24: error: no template named 'span' in namespace 'std'
16 | using SetOfU32 = std::set<uint32_t>;
17 | #if __has_include(<span>)
18 | using SpanOfU32 = std::span<uint32_t>;
   |                        `- error: no template named 'span' in namespace 'std'
19 | #endif
20 | 
```
https://ci.swift.org/job/swift-PR-toolchain-macos/1474/

**Comment by egorzhdan** [comment]
Looks like `-std=c++20` isn't passed in some configurations.
